### ISSUE TITLE
pysidecaster.py GUI unresponsive, no image

### ISSUE THREAD

**Issue by yuanagain** [clariusdev/cast#61]
Hi, when attempting to run:
```
(venv) ➜  clarius-cast-v910-macos git:(master) ✗ LD_LIBRARY_PATH=$LD_LIBRARY_PATH:. python pysidecaster.py
```

I receive a slew of warnings on Qt plugins of the form:
```
...
objc[11487]: Class QDarkNSBox is implemented in both /Users/.../cast/clarius-cast-v910-macos/libcast.dylib (0x10fbaa170) and /Users/.../cast/venv/lib/python3.8/site-packages/PySide2/Qt/plugins/styles/libqmacstyle.dylib (0x126a85dc0). One of the two will be used. Which one is undefined.
...
```

Though the GUI opens, it is unresponsive. See video below:

https://user-images.githubusercontent.com/6503523/166154957-00ef55cb-e974-47fc-95e2-ea4119e16e7a.mov

Any help or guidance would be appreciated--searching through the closed issues I wasn't able to find anything similar.

**Comment by ChristophG123T** [comment]
Yes, I'm getting the same issue. I used the Casting Port given on the Clarius app, and got it to recognize when I was scanning my arm, so I think it's connected. The image does not display, however:
![image](https://user-images.githubusercontent.com/38470529/170783308-0598ba2b-54b3-4caa-8554-036532eccbc5.png)

**Comment by clariusk** [comment]
check your windows firewall, it's likely blocking the port

**Comment by ChristophG123T** [comment]
I have tried granting access to the firewall using a variety of methods, but none have worked. I have tried making it a TCP and UDP port with port number 32947 (the Casting port on the Clarius app and the one that can detect an image being broadcast), and the default 5828 (does not do anything). I have also tried this on the Local and Remote ports.
Right now I have this configured:
![clarError](https://user-images.githubusercontent.com/38470529/170786516-9422d8da-75bd-49b1-b2dc-d4e37081e0a5.jpg)

**Comment by clariusk** [comment]
yes, the port that needs to be unblocked is a udp port which is exposed through the C API
your best bet is to put a rule in for the executable itself, but using python i suppose it's not that simple (some others recently asked about this). so for now, suggest to close your network and disable firewall to get some testing going.

**Comment by yuanagain** [comment]
> check your windows firewall, it's likely blocking the port

I'm on OS X, not on Windows

**Comment by yuanagain** [comment]
@clariusk Can you advise on whether the UI should be responsive? I'm unable to even edit the IP or port fields

**Comment by clariusk** [comment]
you can fix the TCP port (ie. 5828 as default for researchers), but the UDP port used for streaming is dynamically allocated by the OS, so you cannot put a rule for a specific port, but need to allow the executable itself - which as mentioned above it difficult since python is the executable and interpreting the script, there might be ways around this, but quick searching didn't yield anything useful

**Comment by clariusk** [comment]
it may be related to PySide, I'm not sure as I'm not running macOS, you could try the python program without PySide and see if that works

**Comment by julien-l** [comment]
Hello, 9.4.0 is scheduled to release June 16th. It will be a "Pilot" only (it will not be released to the App/Play Store). https://support.clarius.com/hc/en-us/articles/360019548852-Joining-the-Pilot-Program

**Comment by ChristophG123T** [comment]
Hi Kris @clariusk ,

I'm trying to get the latest version of pysidecaster working, but I'm still unable to stream the image or connect to the Clarius' IP. It gives the following error:
![image](https://user-images.githubusercontent.com/38470529/175993785-c72eecef-78ee-4621-ab1a-de17815cc156.png)

It should be noted that it can detect whether an image is streaming or not after I "connect" the Clarius, but will not stream the image nor seem to connect to the IP.
The outgoing UDP port 5828 is not being blocked on my Lab's firewall, so I don't think this is the issue. I also have the newest binaries (9.4) installed, with RF-Mode enabled, and the PC connected to the Clarius network. Lastly, the configuration I used for the connection on PyCharm is this:
![image](https://user-images.githubusercontent.com/38470529/175994926-780ce771-1a72-4aa5-848b-6d58353d7023.png)

I know you mentioned putting a rule in the executable, but I don't know how to do this. If you suspect this is the issue, how can I write this rule?

Thanks,
Chris

**Comment by yuanagain** [comment]
> 9.4 Clarius App pilot and corresponding Cast pre-release are now published

Thank you!

**Comment by yuanagain** [comment]
> https://support.clarius.com/hc/en-us/articles/360019548852-Joining-the-Pilot-Program

Thanks Julien; how does one get access to this pilot?

**Comment by ChristophG123T** [comment]
Hello, I am still having issues getting the Python Clarius demo to work on Windows. How can I create an exception for the executable? Blocking my firewall doesn't seem to work.

**Comment by clariusk** [comment]
just to confirm, in your last message you said you updated the cast binaries, and you also updated to the Clarius App 9.4 ? If the App is not updated, you will not be able to connect either. Once confirmed, and you get to the state where is says imaging running, then if you still see no image, it will be the firewall that you need to completely unblock or I believe another user had success in just unblocking the python executable.

**Comment by clariusk** [comment]
you can follow the install instructions here: https://support.clarius.com/hc/en-us/articles/360019548852-Joining-the-Pilot-Program

**Comment by ChristophG123T** [comment]
I just left and rejoined the Pilot program. When I go to the play store, it confirms that I'm a beta tester of the Clarius app, but it has not given me any updates. Perhaps I should uninstall and reinstall the Clarius app? It still says it's Version 9.3.1 and has not given me any indication of a new update.

**Comment by julien-l** [comment]
It is unclear why Google Play serves the 9.3.1 binaries instead of 9.4.0 in the the beta track. I suspect it is because 9.3.1 was released after 9.4.0 (it's a hot fix) and Google Play serves the more recent one.

As a workaround, you can get the 9.4.0 APK in another repo: https://github.com/clariusdev/mobileapi/releases/tag/v9.4.0.

**Comment by ChristophG123T** [comment]
Ok, the new app is working better, but still doesn't display the image. I am able to connect the Probe without any errors and remotely control the depth (etc.) from my PC. I tried allowing the Clarius port through the firewall (shown in my previous posts) but it still doesn't show images. How can I put a rule in the executable?

**Comment by julien-l** [comment]
Create a rule in the Windows firewall, not in your executable.

**Comment by ChristophG123T** [comment]
I did create a rule in the Windows firewall and it does not stream the image, still:
![image](https://user-images.githubusercontent.com/38470529/178037949-b8b3ee06-62ad-4db3-a97d-f0df76f6ab5b.png)

**Comment by ChristophG123T** [comment]
Ok, so I've been trying to get it working on Windows first, but my long-term goal is to get the Python demo working on my Nvidia Jetson Xavier computer which uses Ubuntu and has an arm64 architecture, so I was thinking it might work better on Ubuntu. However, the binaries you gave are not compatible with the arm64 architecture. Can you build the 9.4.0 binaries for arm64? Thanks

**Comment by julien-l** [comment]
We have no support for Linux arm64 currently, I have logged a feature request it will be evaluated at a later point.

**Comment by ChristophG123T** [comment]
> We have no support for Linux arm64 currently, I have logged a feature request it will be evaluated at a later point.

Ok, thank you! Do you have any idea of when this feature to support arm64 might be available? Everything else needed for the demo seems to be working (we built PySide2 in arm64 and have PyCharm working).

**Comment by clariusk** [comment]
thanks for feedback, both macOS M1/M2 and Linux arm64 are current API platform requests, we have no plans as of yet to support, since our build system is already targeting around 12 platforms including all the variants of Android we need to support. We'll keep in mind should we have time to invest in new builders.

**Comment by ChristophG123T** [comment]
Ok I see. I'll give you all the specs of my software for the Windows demo to try getting it working again:
-Windows x64
-Python 3.8
-Clarius 9.4.0
-PySide2 (latest version)

I have tried using Python 3.9 as an interpreter, but it cannot load the pyclariuscast module:
![image](https://user-images.githubusercontent.com/38470529/178326800-06802356-0d11-4d4e-b6cb-c50368ef74be.png)

 It runs on 3.8 with all other functions (depth changes etc), but does not show the image. I've allowed the port to go through, but it still doesn't show images. The one other solution I haven't tried is putting a rule in the executable, but this seems like it would do the same thing. Lastly, I'm using the latest version of PySide2. Maybe an older version of PySide2 would work?

**Comment by clariusk** [comment]
yes, our builders are on python 3.8, so it needs to match up (i still don't know why python breaks ABI compatibility, i would need to investigate this more: https://docs.python.org/3/c-api/stable.html)

for the firewall, the port you connect to and provide in your code is NOT the port you need to unblock, as that is the TCP port. there is another UDP port opened in the background used for streaming, and it's not fixed, hence you need to make a rule for your whole executable (in this case python.exe)

**Comment by clariusk** [comment]
https://docs.microsoft.com/en-us/windows/security/threat-protection/windows-firewall/create-an-inbound-port-rule shows how to set a rule for a program.

[cusCastConnect](https://github.com/clariusdev/cast/blob/master/src/include/cast/cast.h) returns the port through the callback, but this is not part of the python libs, so unblocking whole program will work

**Comment by ChristophG123T** [comment]
> https://docs.microsoft.com/en-us/windows/security/threat-protection/windows-firewall/create-an-inbound-port-rule shows how to set a rule for a program.
> 
> [cusCastConnect](https://github.com/clariusdev/cast/blob/master/src/include/cast/cast.h) returns the port through the callback, but this is not part of the python libs, so unblocking whole program will work

Ok, I have made both an outbound and inbound rule this way, but it still does not show images.
First, I created a new custom rule, and applied it to the Python 3.8 python.exe:
![image](https://user-images.githubusercontent.com/38470529/178339523-03d2e19d-4275-47a0-9ac6-31e113c331a7.png)
Then I allowed any UDP ports to go through (you said this changes):
![image](https://user-images.githubusercontent.com/38470529/178339628-9e095897-986b-4268-8a3a-8de001d24291.png)
Then, I let any IP address through:
![image](https://user-images.githubusercontent.com/38470529/178339672-6bc1fdcf-a050-4153-a029-32a62dfa5d8f.png)
And allowed the connection:
![image](https://user-images.githubusercontent.com/38470529/178339719-e07223cb-634b-44df-ad46-f625c0ad52c8.png)
Applied it to every domain:
![image](https://user-images.githubusercontent.com/38470529/178339769-c72fa43c-d848-4831-8f9b-280319e6f21e.png)
And then finished.
![image](https://user-images.githubusercontent.com/38470529/178339835-ab8d28ce-ff46-42c3-8617-b03df80fbff5.png)

Was there any issues in this setup that I can edit?

**Comment by clariusk** [comment]
I believe in similar recent threads @geisjb01 and @yuanagain have been able to get things running, they may have some insights on what rules they put in place.

**Comment by dgblack** [comment]
Not sure if this is still helpful, but I had the exact same problem with the C++ demos. Looking at the firewall rule on the list of active rules, I noticed that %USERPROFILE% evaluates to something weird and wrong. Changing that program path to the actual explicit path of my .exe file (i.e. C:\users\...\caster_qt.exe) fixed it for me. The streaming works well now. Note, I only needed an inbound rule, not outbound, and I set the protocol type to Any.
